### ISSUE TITLE
Docs for Cooperator and friends don't say what they are for

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6213]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#6213|
|Type|enhancement|
|Created|2012-12-20 19:13:03Z|
|Branch|https://github.com/twisted/twisted/tree/cooperator-docs-6213|

The API docs for `Cooperator`, `coiterate` and `cooperate` do an excellent job at explaining how to use them, but don't cover why anyone would want to use them.  There's also no such documentation in the howtos.

Attachments:

* [cooperator-docs-6213.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f4f/f4f5524a58df05dbc5153b43dab74f3a749346f7/1bb3c6faf34f876463069b926acc19fc17b7a63b.patch) (2546 bytes) - added by jml on 2013-08-22 03:02:24Z - 
* [cooperator-docs-6213-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f4f/f4f5524a58df05dbc5153b43dab74f3a749346f7/da8e7d259a7d8a76cf9c461d12116920bb32d3ec.patch) (2546 bytes) - added by jml on 2013-08-22 13:07:52Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6213 6213
type__enhancement enhancement
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_cooperator_docs_6213 branches/cooperator-docs-6213
branch_author__jml jml
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1356030783000000 1356030783000000
changetime__1474198049187875 1474198049187875
version__None None
owner__jml jml

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml<br>@jml set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml removed owner<br>@jml set status to `new`|
|-|-|

Sorry for patch vs branch. Don't have keys on this laptop.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_3"></a>|@jonathanj commented|
|-|-|

It's always good to see the documentation improve, and in particular something like Cooperator.

Not a review but I have a comment: The documentation hints at an answer to the ever popular question of limiting parallelism but doesn't go as far as to explain how you would do it or how it works. However, documenting this aspect of it seems better suited to a howto than a docstring and it would be great to have this useful bit of knowledge in the Twisted documentation.

Also, there's a small grammar error: "it distribute work between all tasks."

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

Thanks for the comments.

I filed [#6693](https://github.com/twisted/twisted/issues/6693), since I think that the right way to deal with the 'limited parallelism' problem is to provide a helper, and that such is beyond the scope of this ticket. 

Good catch on the grammar mistake. Will upload a fix.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

(In [39817]) Branching to 'cooperator-docs-6213'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

(In [39819]) Apply Alex_Gaynor's suggestions (refs [#6213](https://github.com/twisted/twisted/issues/6213))

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_8"></a>|@dreid set owner to @jml|
|-|-|

Hey jml,

This new description is great, however when explaining the difference between `coiterate` and `cooperate` you only really talk about `coiterate` it might be helpful to point out that `cooperate` returns an object that can be paused, resumed, and stopped for greater control over the execution of the task.  

And a few formatting issues.

 * The opening and the closing of the docstrings should be on a line by themselves.
   https://github.com/twisted/twisted/compare/trunk...cooperator-docs-6213#L1R500
 * We appear to use `L{}` even when referring to the class that is currently being documented (based on documentation from `twisted.web.client.Response`).  So `C{Cooperator}` should probably be `L{Cooperator}}`.
 * `    L{coiterate}.  They are equivalent, but C{coiterate} returns a`  Second reference to coiterate should use `L{}`.

Once these are addressed you should force a build so we can make sure no new pydoctor errors were introduced.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

```
"""Cooperative task scheduler.
```

This is not the docstring style used by Twisted.  The old style:

```
"""
Cooperative task scheduler.
```

was correct.

> This new description is great, however when explaining the difference between coiterate and cooperate you only really talk about coiterate it might be helpful to point out that cooperate returns an object that can be paused, resumed, and stopped for greater control over the execution of the task.

Indeed.  I think `cooperate` should be encouraged and `coiterate` should be discouraged.  `cooperate` is strictly more functional and a lot of people seem to get confused by the word `coiterate`.

Also, for `Cooperator` to achieve maximum effect, the global instance should be used in most cases.  Multiple `Cooperator`s do not cooperate with each other!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml commented|
|-|-|

(In [39824]) Correctly formatted docstring (refs [#6213](https://github.com/twisted/twisted/issues/6213))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml commented|
|-|-|

(In [39825]) Always link to methods & classes, rather than just the first time (refs [#6213](https://github.com/twisted/twisted/issues/6213))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml commented|
|-|-|

(In [39827]) Recommend `cooperate` over `coiterate` (refs [#6213](https://github.com/twisted/twisted/issues/6213))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_13"></a>|@jml commented|
|-|-|

(In [39830]) Recommend global cooperator (refs [#6213](https://github.com/twisted/twisted/issues/6213))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml removed owner|
|-|-|

Thanks for the feedback!

dreid:
 * have outlined `cooperate` more
 * fixed docstring formatting
 * link everywhere (fwiw, I was following the practice used by wikipedia & others of only linking the first reference)

exarkun:
 * fixed docstring formatting
 * have recommended `cooperate` over `coiterate` (I thought it was the other way around, and that `CooperativeTask` existed more for internal use only. Glad to learn otherwise.)
 * have recommended the global cooperator, specifically by recommending `task.cooperate`.

I've changed the ordering of the paragraphs and made a couple of spelling fixes along the way.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj set owner to @jml<br>@wallrj set status to `new`|
|-|-|

Thanks jml.

The new docstrings are very helpful.

Notes:
 * Forced Build
   * https://buildbot.twistedmatrix.com/builds/apidocs/apidocs-157220fe4f0afd94944130c844c49944acacb3c6.tar.bz2

Points:
 1. Some sentences which confused me, or which I thought could be
    clearer (and my suggested changes):
    1. "then work will only resume after it fires" > "then the
       Cooperator will pause until the deferred fires"

    2. "doing one thing, waiting for a Deferred to fire, doing the
       next thing, repeat " > "serializing asynchronous tasks"
       1. but isn't that just a deferred callback chain?

    3. "Multiple Cooperator's do not" > "Cooperators do not"
       1. don't think the apostrophe is necessary.
       2. also what exactly does this mean?
          1. Does it mean that you *can* have multiple cooperators but
             you shouldn't need to unless...eg you create a Cooperator
             subclass with a special scheduling policy.

 2. If we want to encourage the use of the global cooperater then I'd
    have thought you should move (or copy) this documentation to that
    function (task.cooperate) so people see it clearly in the task API
    index and visit it before delving into task.Cooperator.

 3. Also it might be even clearer if task.cooperate was moved above
    task.Cooperate in the source code...which I think will cause
    pydoctor to present them in that order in the API docs.

 4. The ticket description also promises some howto documentation -
    which I think is badly needed.
    1. If you don't want to add it here please create a followup
       ticket for it.

    2. http://as.ynchrono.us/2006/05/limiting-parallelism_22.html

       1. That's exarkun's document that I've referred to in the
          past, but in it he's not using the global cooperate (maybe
          it didn't exist at the time)

 5. Is there a good external description of the pattern that
    Cooperate implements? If so, it would be nice to link to it

    1. Maybe this? https://en.wikipedia.org/wiki/Computer_multitasking#Cooperative_multitasking.2Ftime-sharing

I think it's a great improvement anyway.

So +1 from me if you address or answer the numbered points above. Or
resubmit for another review if you want a another opinion.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_18"></a>|@wallrj commented|
|-|-|

Hmm. My carefully numbered list got mangled...but hopefully you get the point.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj commented|
|-|-|

Some more cooperator descriptions which might be re-used in a howto:
 * http://stackoverflow.com/questions/2861858/queue-remote-calls-to-a-python-twisted-perspective-broker/2862440#2862440
 * http://technicae.cogitat.io/2008/06/async-batching-with-twisted-walkthrough.html
 * http://blog.mailgun.com/post/stress-testing-http-with-twisted-python-and-treq/

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_20"></a>|@jml commented|
|-|-|

(In [39855]) Respond to rwall's review comments (refs [#6213](https://github.com/twisted/twisted/issues/6213))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_21"></a>|@jml commented|
|-|-|

(In [39856]) Add some 'why' documentation to `cooperate` (refs [#6213](https://github.com/twisted/twisted/issues/6213))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_22"></a>|@jml commented|
|-|-|

Thanks Richard.

1. Fixed

2. Supplemented, rather than replaced.

3. Apostrophe fixed, no clarifying text added. See below.

2. I believe the documentation for the global cooperator is adequately suited to its purpose. See below.

3. Interesting notion. The layout as-is makes the code very clear, so I'd prefer to keep it this way.

4. Howto is definitely out of scope. The comment in the original ticket was there to emphasise that I'd checked everywhere, rather than to suggest that all of these places should have documentation. Nevertheless have filed [#6724](https://github.com/twisted/twisted/issues/6724).

2. That's because the global cooperator is not appropriate for that use case. As mentioned earlier in the thread, I think the right way to solve the limited parallelism confusion is just to add the helper (see [#6693](https://github.com/twisted/twisted/issues/6693))

5. I'd defer to exarkun on that one, and would rather omit a potentially useful link than include a potentially misleading one.


In general, the more I learn about cooperator, the more it seems that most users will not need to use it directly: they'll want `cooperate` and eventually `parallel`. Thus it's appropriate for the docstring of the class to go into a bit of detail about its power and the approach it takes, and likewise appropriate for the global `cooperate` function to have high-level documentation, focusing on its most common use case.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_23"></a>|@wallrj commented|
|-|-|

Ok. I scanned through the changes and noticed one mistake.

 * r39856 "Just break each task up into so" -- there's something missing.

Fix that, add a .doc news file and please merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_24"></a>|@jml commented|
|-|-|

(In [39857]) Fix thinko in cooperate docstring (refs [#6213](https://github.com/twisted/twisted/issues/6213))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_25"></a>|@jml set status to `closed`|
|-|-|

(In [39860]) API documentation for twisted.internet.task.Cooperator

 * Author: jml
 * Reviewers: Alex_Gaynor, dreid, exarkun, rwall
 * Fixes: [#6213](https://github.com/twisted/twisted/issues/6213)

Extends existing API documentation by providing some context on the class
as a whole and suggesting some reasons to use it.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_26"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
